File: games/mini_royal.py
import asyncio
import random
import math
import json
from datetime import datetime
from enum import Enum
from typing import Dict, List, Optional

class GameState(Enum):
    WAITING = "waiting"
    COUNTDOWN = "countdown"
    ACTIVE = "active"
    ENDED = "ended"

class Player:
    def __init__(self, user_id: int, username: str, skin: str = None, weapon: str = None):
        self.user_id = user_id
        self.username = username
        self.skin = skin or "🧍"  # Default standing emoji
        self.weapon = weapon
        self.position = (random.uniform(0.2, 0.8), random.uniform(0.2, 0.8))
        self.direction = random.uniform(0, 2 * math.pi)
        self.speed = 0.005
        self.alive = True
        self.last_shot = 0
        self.shot_cooldown = 0.5
        self.kills = 0
        self.state = "idle"
        self.animation_frame = 0

class MiniRoyalGame:
    def __init__(self, game_id: str, max_players: int = 10, map_id: str = "classic"):
        self.game_id = game_id
        self.max_players = max_players
        self.players: Dict[int, Player] = {}
        self.state = GameState.WAITING
        self.start_time = None
        self.duration = 90
        self.circle_radius = 1.0
        self.circle_shrink_rate = 0.001
        self.circle_final_radius = 0.2
        self.bullets = []
        self.spectators = set()
        self.map_id = map_id
        self.animation_time = 0
        
        # Animation states
        self.animation_states = {
            "idle": ["🧍", "🧍", "🧍", "🧍", "🧍", "🧍", "🧍", "🧍"],
            "walking": ["🚶", "🚶", "🚶‍➡️", "🚶‍➡️", "🚶", "🚶", "🚶‍⬅️", "🚶‍⬅️"],
            "running": ["🏃", "🏃", "🏃‍➡️", "🏃‍➡️", "🏃", "🏃", "🏃‍⬅️", "🏃‍⬅️"],
            "shooting": ["🧍", "🧍", "🔫", "🔫", "🧍", "🧍", "🧍", "🧍"],
            "dead": ["💀", "💀", "💀", "💀", "💀", "💀", "💀", "💀"]
        }
        
        # Available skins
        self.skin_categories = {
            "basic": ["😎", "🤓", "🧐", "😤", "🤬", "😺", "😼", "😾", "🙉", "🫩", "🤕"],
            "advanced": ["🥷", "👮", "🧑‍✈️", "🫅", "🧛", "🎅", "🦹", "🦸", "🕵️"],
            "animal": ["🦍", "🐯", "🦁", "🐼", "🐨", "🦂", "🐵", "🐶", "🐱"],
            "legendary": ["👾", "🤖", "🦄", "🐲", "🦅", "🦸‍♂️", "🦸‍♀️", "🧙‍♂️", "🧙‍♀️"]
        }
        
        # Available weapons
        self.weapons = {
            "pistol": {"emoji": "🔫", "damage": 10, "cooldown": 0.5},
            "knife": {"emoji": "🔪", "damage": 15, "cooldown": 0.3},
            "bow": {"emoji": "🏹", "damage": 20, "cooldown": 0.7},
            "laser": {"emoji": "⚡", "damage": 25, "cooldown": 1.0}
        }
        
        # Map configurations
        self.maps = {
            "classic": {
                "name": "Classic Arena",
                "backgroundColor": "#000000",
                "safeZoneColor": "#00ff00",
                "dangerZoneColor": "rgba(255, 0, 0, 0.3)",
                "features": ["grid"]
            },
            "desert": {
                "name": "Desert Dunes",
                "backgroundColor": "#EDC9AF",
                "safeZoneColor": "#FFD700",
                "dangerZoneColor": "rgba(139, 69, 19, 0.4)",
                "features": ["dunes"]
            },
            "arctic": {
                "name": "Frozen Tundra",
                "backgroundColor": "#F0F8FF",
                "safeZoneColor": "#00BFFF",
                "dangerZoneColor": "rgba(176, 224, 230, 0.4)",
                "features": ["snowflakes", "ice_cracks"]
            }
        }
        
    def add_player(self, user_id: int, username: str, skin: str = None, weapon: str = None) -> bool:
        if len(self.players) >= self.max_players or self.state != GameState.WAITING:
            return False
            
        self.players[user_id] = Player(user_id, username, skin, weapon)
        return True
        
    def remove_player(self, user_id: int):
        if user_id in self.players:
            del self.players[user_id]
            
    def start_game(self):
        if len(self.players) < 2:
            return False
            
        self.state = GameState.COUNTDOWN
        self.start_time = datetime.now()
        return True
        
    async def game_loop(self):
        # Countdown
        for i in range(3, 0, -1):
            await self.broadcast({
                "type": "countdown", 
                "value": i,
                "map": self.maps[self.map_id]
            })
            await asyncio.sleep(1)
            
        self.state = GameState.ACTIVE
        start_time = datetime.now()
        self.animation_time = datetime.now().timestamp()
        
        # Main game loop
        while self.state == GameState.ACTIVE:
            current_time = datetime.now()
            elapsed = (current_time - start_time).total_seconds()
            remaining = self.duration - elapsed
            
            # Update animation time
            new_animation_time = current_time.timestamp()
            animation_delta = new_animation_time - self.animation_time
            self.animation_time = new_animation_time
            
            if remaining <= 0:
                self.state = GameState.ENDED
                break
                
            # Update safe zone
            shrink_progress = elapsed / self.duration
            self.circle_radius = self.circle_radius - (
                (self.circle_radius - self.circle_final_radius) * shrink_progress
            )
            
            # Update player positions and states
            for player in self.players.values():
                if not player.alive:
                    player.state = "dead"
                    continue
                    
                # Update animation frame
                player.animation_frame = (player.animation_frame + 1) % 8
                
                # Determine player state
                player.state = "idle"
                
                # Move player
                new_x = player.position[0] + math.cos(player.direction) * player.speed
                new_y = player.position[1] + math.sin(player.direction) * player.speed
                
                # Bounce off walls
                if new_x < 0 or new_x > 1:
                    player.direction = math.pi - player.direction
                    new_x = max(0, min(1, new_x))
                    player.state = "walking"
                if new_y < 0 or new_y > 1:
                    player.direction = -player.direction
                    new_y = max(0, min(1, new_y))
                    player.state = "walking"
                    
                player.position = (new_x, new_y)
                
                # Check if outside safe zone
                center = (0.5, 0.5)
                distance = math.sqrt((new_x - center[0])**2 + (new_y - center[1])**2)
                if distance > self.circle_radius:
                    # Take damage outside zone
                    player.alive = False
                    await self.broadcast({
                        "type": "player_eliminated", 
                        "user_id": player.user_id,
                        "reason": "zone"
                    })
            
            # Update bullets
            new_bullets = []
            for bullet in self.bullets:
                bullet_lifetime = 2.0
                if datetime.now().timestamp() - bullet["time"] < bullet_lifetime:
                    # Check for collisions
                    for player in self.players.values():
                        if (player.alive and player.user_id != bullet["shooter_id"] and
                            self.distance(bullet["position"], player.position) < 0.03):
                            player.alive = False
                            # Award kill to shooter
                            if bullet["shooter_id"] in self.players:
                                self.players[bullet["shooter_id"]].kills += 1
                            await self.broadcast({
                                "type": "player_eliminated", 
                                "user_id": player.user_id,
                                "reason": "shot",
                                "killer_id": bullet["shooter_id"]
                            })
                            break
                    else:
                        new_bullets.append(bullet)
            
            self.bullets = new_bullets
            
            # Check win condition
            alive_players = [p for p in self.players.values() if p.alive]
            if len(alive_players) <= 1:
                self.state = GameState.ENDED
                winner = alive_players[0] if alive_players else None
                await self.broadcast({
                    "type": "game_end",
                    "winner_id": winner.user_id if winner else None,
                    "players": {p.user_id: {
                        "kills": p.kills, 
                        "alive": p.alive,
                        "skin": p.skin,
                        "weapon": p.weapon
                    } for p in self.players.values()}
                })
                break
                
            # Send game state to all players
            await self.broadcast({
                "type": "game_state",
                "players": {user_id: {
                    "position": player.position,
                    "alive": player.alive,
                    "skin": player.skin,
                    "weapon": player.weapon,
                    "state": player.state,
                    "animation_frame": player.animation_frame,
                    "direction": player.direction
                } for user_id, player in self.players.items()},
                "bullets": [{
                    "position": b["position"],
                    "direction": b["direction"],
                    "shooter_id": b["shooter_id"]
                } for b in self.bullets],
                "circle_radius": self.circle_radius,
                "time_remaining": remaining,
                "map": self.maps[self.map_id],
                "animation_time": self.animation_time
            })
            
            await asyncio.sleep(0.05)
            
    async def broadcast(self, data):
        # Implementation depends on your WebSocket setup
        # This would send data to all connected clients
        pass
        
    def player_shoot(self, user_id: int, direction: float):
        if user_id not in self.players or not self.players[user_id].alive:
            return False
            
        player = self.players[user_id]
        current_time = datetime.now().timestamp()
        
        if current_time - player.last_shot < player.shot_cooldown:
            return False
            
        player.last_shot = current_time
        player.state = "shooting"
        
        self.bullets.append({
            "shooter_id": user_id,
            "position": player.position,
            "direction": direction,
            "time": current_time
        })
        
        return True
        
    def distance(self, pos1, pos2):
        return math.sqrt((pos1[0] - pos2[0])**2 + (pos1[1] - pos2[1])**2)
    
    def change_map(self, map_id: str):
        if map_id in self.maps:
            self.map_id = map_id
            return True
        return False

File: games/static/mini-royal/mini-royal.js
class MiniRoyalGame {
    constructor() {
        this.canvas = document.getElementById('game-canvas');
        this.ctx = this.canvas.getContext('2d');
        this.socket = null;
        this.gameId = null;
        this.userId = null;
        this.username = null;
        this.gameState = {
            players: {},
            bullets: [],
            circle_radius: 1.0,
            time_remaining: 90,
            state: 'waiting',
            map: null,
            animation_time: 0
        };
        
        this.animationStates = {
            idle: ["🧍", "🧍", "🧍", "🧍", "🧍", "🧍", "🧍", "🧍"],
            walking: ["🚶", "🚶", "🚶‍➡️", "🚶‍➡️", "🚶", "🚶", "🚶‍⬅️", "🚶‍⬅️"],
            running: ["🏃", "🏃", "🏃‍➡️", "🏃‍➡️", "🏃", "🏃", "🏃‍⬅️", "🏃‍⬅️"],
            shooting: ["🧍", "🧍", "🔫", "🔫", "🧍", "🧍", "🧍", "🧍"],
            dead: ["💀", "💀", "💀", "💀", "💀", "💀", "💀", "💀"]
        };
        
        this.weaponOffsets = {
            pistol: { x: 15, y: -10 },
            knife: { x: 10, y: -5 },
            bow: { x: 20, y: -15 },
            laser: { x: 15, y: -20 }
        };
        
        this.animationFrame = 0;
        this.lastAnimationTime = 0;
        this.animationSpeed = 200;
        this.scale = Math.min(this.canvas.width, this.canvas.height);
        
        this.init();
    }

    init() {
        const urlParams = new URLSearchParams(window.location.search);
        this.gameId = urlParams.get('game_id') || this.generateGameId();
        this.userId = this.getUserId();
        this.username = this.getUsername();
        
        if (!urlParams.get('game_id')) {
            window.history.replaceState({}, '', `${window.location.pathname}?game_id=${this.gameId}`);
        }

        this.connectWebSocket();
        this.setupEventListeners();
        this.resizeCanvas();
        
        this.render();
    }

    generateGameId() {
        return 'game_' + Math.random().toString(36).substr(2, 9);
    }

    getUserId() {
        return Math.floor(Math.random() * 1000000).toString();
    }
    
    getUsername() {
        return 'Player_' + Math.floor(Math.random() * 1000);
    }

    connectWebSocket() {
        const protocol = window.location.protocol === 'https:' ? 'wss:' : 'ws:';
        const wsUrl = `${protocol}//${window.location.host}/ws/mini-royal/${this.gameId}`;
        
        this.socket = new WebSocket(wsUrl);
        
        this.socket.onopen = () => {
            this.socket.send(JSON.stringify({
                type: 'join',
                user_id: this.userId,
                username: this.username,
                game_id: this.gameId
            }));
        };
        
        this.socket.onmessage = (event) => {
            const data = JSON.parse(event.data);
            this.handleMessage(data);
        };
        
        this.socket.onclose = () => {
            setTimeout(() => this.connectWebSocket(), 3000);
        };
        
        this.socket.onerror = (error) => {
            console.error('WebSocket error:', error);
        };
    }

    handleMessage(data) {
        switch (data.type) {
            case 'game_state':
                this.gameState = data;
                this.updatePlayerCount();
                break;
                
            case 'countdown':
                this.showCountdown(data.value);
                this.gameState.map = data.map;
                break;
                
            case 'player_joined':
                this.addPlayer(data.player);
                break;
                
            case 'player_left':
                this.removePlayer(data.user_id);
                break;
                
            case 'player_eliminated':
                this.handlePlayerEliminated(data.user_id, data.reason, data.killer_id);
                break;
                
            case 'game_start':
                this.startGame();
                break;
                
            case 'game_end':
                this.endGame(data.winner_id, data.players);
                break;
                
            case 'error':
                console.error('Game error:', data.message);
                break;
        }
    }

    setupEventListeners() {
        const shootBtn = document.getElementById('shoot-btn');
        shootBtn.addEventListener('click', () => {
            const centerX = 0.5;
            const centerY = 0.5;
            const player = this.gameState.players[this.userId];
            if (!player) return;
            
            const dx = centerX - player.position[0];
            const dy = centerY - player.position[1];
            const direction = Math.atan2(dy, dx);
            
            this.socket.send(JSON.stringify({
                type: 'shoot',
                direction: direction
            }));
        });

        const startBtn = document.getElementById('start-btn');
        startBtn.addEventListener('click', () => {
            this.socket.send(JSON.stringify({ type: 'start_game' }));
        });

        const playAgainBtn = document.getElementById('play-again');
        playAgainBtn.addEventListener('click', () => {
            window.location.reload();
        });

        const backToLobbyBtn = document.getElementById('back-to-lobby');
        backToLobbyBtn.addEventListener('click', () => {
            window.location.href = '/games';
        });

        window.addEventListener('resize', () => {
            this.resizeCanvas();
        });
    }

    resizeCanvas() {
        const container = this.canvas.parentElement;
        const size = Math.min(container.clientWidth, container.clientHeight) * 0.9;
        
        this.canvas.width = size;
        this.canvas.height = size;
        this.scale = size;
    }

    render() {
        requestAnimationFrame(() => this.render());
        
        this.ctx.clearRect(0, 0, this.canvas.width, this.canvas.height);
        
        this.drawBackground();
        this.drawSafeZone();
        this.drawPlayers();
        this.drawBullets();
        this.drawUI();
    }

    drawBackground() {
        if (!this.gameState.map) {
            this.ctx.fillStyle = '#000';
            this.ctx.fillRect(0, 0, this.canvas.width, this.canvas.height);
            return;
        }
        
        this.ctx.fillStyle = this.gameState.map.backgroundColor;
        this.ctx.fillRect(0, 0, this.canvas.width, this.canvas.height);
        
        if (this.gameState.map.features.includes('grid')) {
            this.drawGrid();
        }
        
        if (this.gameState.map.features.includes('dunes')) {
            this.drawDunes();
        }
        
        if (this.gameState.map.features.includes('snowflakes')) {
            this.drawSnowflakes();
        }
        
        if (this.gameState.map.features.includes('ice_cracks')) {
            this.drawIceCracks();
        }
    }
    
    drawGrid() {
        const size = 20;
        this.ctx.strokeStyle = 'rgba(255, 255, 0, 0.1)';
        this.ctx.lineWidth = 1;
        
        for (let x = 0; x <= this.canvas.width; x += size) {
            this.ctx.beginPath();
            this.ctx.moveTo(x, 0);
            this.ctx.lineTo(x, this.canvas.height);
            this.ctx.stroke();
        }
        
        for (let y = 0; y <= this.canvas.height; y += size) {
            this.ctx.beginPath();
            this.ctx.moveTo(0, y);
            this.ctx.lineTo(this.canvas.width, y);
            this.ctx.stroke();
        }
    }
    
    drawDunes() {
        this.ctx.fillStyle = '#D2B48C';
        for (let i = 0; i < 15; i++) {
            const x = Math.random() * this.canvas.width;
            const y = Math.random() * this.canvas.height;
            const width = 30 + Math.random() * 40;
            const height = 10 + Math.random() * 15;
            
            this.ctx.beginPath();
            this.ctx.moveTo(x, y);
            this.ctx.quadraticCurveTo(x + width/2, y - height, x + width, y);
            this.ctx.fill();
        }
    }
    
    drawSnowflakes() {
        this.ctx.fillStyle = '#FFFFFF';
        const time = Date.now() / 1000;
        
        for (let i = 0; i < 50; i++) {
            const x = (Math.sin(time + i) * 0.5 + 0.5) * this.canvas.width;
            const y = (i/50 + time * 0.1) % 1 * this.canvas.height;
            const size = 2 + Math.sin(time + i) * 1;
            
            this.ctx.beginPath();
            this.ctx.arc(x, y, size, 0, Math.PI * 2);
            this.ctx.fill();
        }
    }
    
    drawIceCracks() {
        this.ctx.strokeStyle = 'rgba(173, 216, 230, 0.6)';
        this.ctx.lineWidth = 1;
        
        for (let i = 0; i < 8; i++) {
            const startX = Math.random() * this.canvas.width;
            const startY = Math.random() * this.canvas.height;
            
            this.ctx.beginPath();
            this.ctx.moveTo(startX, startY);
            
            let currentX = startX;
            let currentY = startY;
            const segments = 5 + Math.floor(Math.random() * 5);
            
            for (let s = 0; s < segments; s++) {
                const angle = Math.random() * Math.PI * 2;
                const length = 20 + Math.random() * 30;
                
                currentX += Math.cos(angle) * length;
                currentY += Math.sin(angle) * length;
                
                currentX = Math.max(0, Math.min(this.canvas.width, currentX));
                currentY = Math.max(0, Math.min(this.canvas.height, currentY));
                
                this.ctx.lineTo(currentX, currentY);
            }
            
            this.ctx.stroke();
        }
    }

    drawSafeZone() {
        const centerX = this.canvas.width / 2;
        const centerY = this.canvas.height / 2;
        const radius = this.gameState.circle_radius * this.scale / 2;
        
        if (!this.gameState.map) return;
        
        this.ctx.beginPath();
        this.ctx.arc(centerX, centerY, radius, 0, Math.PI * 2);
        this.ctx.strokeStyle = this.gameState.map.safeZoneColor;
        this.ctx.lineWidth = 3;
        this.ctx.stroke();
        
        this.ctx.fillStyle = this.gameState.map.safeZoneColor.replace(')', ', 0.1)').replace('rgb', 'rgba');
        this.ctx.globalAlpha = 0.1;
        this.ctx.fill();
        this.ctx.globalAlpha = 1.0;
        
        this.drawDangerZone(centerX, centerY, radius);
    }
    
    drawDangerZone(centerX, centerY, safeRadius) {
        if (!this.gameState.map) return;
        
        const maxRadius = Math.max(this.canvas.width, this.canvas.height);
        const gradient = this.ctx.createRadialGradient(
            centerX, centerY, safeRadius,
            centerX, centerY, maxRadius
        );
        gradient.addColorStop(0, 'transparent');
        gradient.addColorStop(0.5, this.gameState.map.dangerZoneColor);
        gradient.addColorStop(1, this.gameState.map.dangerZoneColor);
        
        this.ctx.fillStyle = gradient;
        this.ctx.fillRect(0, 0, this.canvas.width, this.canvas.height);
        
        const pulse = Math.sin(Date.now() / 500) * 0.2 + 0.8;
        this.ctx.globalAlpha = 0.3 * pulse;
        this.ctx.fillStyle = this.gameState.map.dangerZoneColor.replace('0.4', '0.6');
        this.ctx.fillRect(0, 0, this.canvas.width, this.canvas.height);
        this.ctx.globalAlpha = 1.0;
    }

    drawPlayers() {
        for (const userId in this.gameState.players) {
            const player = this.gameState.players[userId];
            if (!player.alive) continue;
            
            const x = player.position[0] * this.canvas.width;
            const y = player.position[1] * this.canvas.height;
            const size = this.scale * 0.05;
            const isCurrentPlayer = userId === this.userId;
            
            this.drawCharacter(x, y, size, player, isCurrentPlayer);
        }
    }
    
    drawCharacter(x, y, size, player, isCurrentPlayer) {
        const state = player.state || 'idle';
        const frame = player.animation_frame || 0;
        
        let emoji;
        if (this.animationStates[state]) {
            emoji = this.animationStates[state][frame % 8];
        } else {
            emoji = player.skin || '🧍';
        }
        
        this.ctx.font = `${size}px Arial`;
        this.ctx.textAlign = 'center';
        this.ctx.textBaseline = 'middle';
        
        if (isCurrentPlayer) {
            this.ctx.shadowBlur = 10;
            this.ctx.shadowColor = '#ff0';
        }
        
        this.ctx.fillText(emoji, x, y);
        
        if (isCurrentPlayer) {
            this.ctx.shadowBlur = 0;
        }
        
        if (player.weapon) {
            this.drawWeapon(x, y, size, player.weapon, state, frame);
        }
        
        this.ctx.fillStyle = '#fff';
        this.ctx.font = '12px Arial';
        this.ctx.textAlign = 'center';
        this.ctx.fillText(player.username || 'Player', x, y - size - 5);
    }
    
    drawWeapon(x, y, size, weapon, state, frame) {
        const weaponEmojis = {
            pistol: "🔫",
            knife: "🔪",
            bow: "🏹",
            laser: "⚡"
        };
        
        const emoji = weaponEmojis[weapon] || "🔫";
        const offset = this.weaponOffsets[weapon] || { x: 15, y: -10 };
        
        const frameOffset = {
            x: offset.x + (Math.sin(frame * 0.5) * 2),
            y: offset.y + (Math.cos(frame * 0.5) * 2)
        };
        
        this.ctx.font = `${size * 0.6}px Arial`;
        this.ctx.fillText(emoji, x + frameOffset.x, y + frameOffset.y);
    }

    drawBullets() {
        for (const bullet of this.gameState.bullets) {
            const x = bullet.position[0] * this.canvas.width;
            const y = bullet.position[1] * this.canvas.height;
            const size = this.scale * 0.01;
            
            this.ctx.beginPath();
            this.ctx.arc(x, y, size, 0, Math.PI * 2);
            this.ctx.fillStyle = '#ff0';
            this.ctx.fill();
            
            const trailLength = 10;
            const trailX = x - Math.cos(bullet.direction) * trailLength;
            const trailY = y - Math.sin(bullet.direction) * trailLength;
            
            this.ctx.beginPath();
            this.ctx.moveTo(trailX, trailY);
            this.ctx.lineTo(x, y);
            this.ctx.strokeStyle = 'rgba(255, 255, 0, 0.5)';
            this.ctx.lineWidth = 2;
            this.ctx.stroke();
        }
    }

    drawUI() {
        const timeEl = document.getElementById('time-remaining');
        if (timeEl) {
            timeEl.textContent = Math.ceil(this.gameState.time_remaining || 0);
        }
        
        const circleProgress = document.querySelector('.circle-progress');
        if (circleProgress && this.gameState.time_remaining) {
            const percentage = (this.gameState.time_remaining / 90) * 100;
            circleProgress.style.background = `conic-gradient(#ff0 ${percentage}%, transparent ${percentage}%)`;
        }
        
        this.updatePlayerCount();
    }

    updatePlayerCount() {
        const aliveCount = Object.values(this.gameState.players).filter(p => p.alive).length;
        const totalCount = Object.keys(this.gameState.players).length;
        
        document.getElementById('alive-count').textContent = aliveCount;
        document.getElementById('total-count').textContent = totalCount;
    }

    showCountdown(value) {
        let countdownEl = document.getElementById('countdown');
        if (!countdownEl) {
            countdownEl = document.createElement('div');
            countdownEl.id = 'countdown';
            countdownEl.style.position = 'absolute';
            countdownEl.style.top = '50%';
            countdownEl.style.left = '50%';
            countdownEl.style.transform = 'translate(-50%, -50%)';
            countdownEl.style.fontSize = '100px';
            countdownEl.style.color = '#ff0';
            countdownEl.style.zIndex = '100';
            document.querySelector('.game-container').appendChild(countdownEl);
        }
        
        countdownEl.textContent = value;
        
        if (value === 0) {
            setTimeout(() => countdownEl.remove(), 1000);
        }
    }

    addPlayer(player) {
        this.gameState.players[player.user_id] = player;
        this.updatePlayerCount();
    }

    removePlayer(userId) {
        delete this.gameState.players[userId];
        this.updatePlayerCount();
    }

    handlePlayerEliminated(userId, reason, killerId) {
        if (userId === this.userId) {
            alert('You were eliminated! ' + (reason === 'zone' ? 'You left the safe zone.' : 'You were shot.'));
        }
        this.removePlayer(userId);
    }

    startGame() {
        document.getElementById('lobby').classList.add('hidden');
        document.getElementById('game-ui').classList.remove('hidden');
        this.gameState.state = 'active';
    }

    endGame(winnerId, players) {
        this.gameState.state = 'ended';
        
        const gameOverEl = document.getElementById('game-over');
        gameOverEl.classList.remove('hidden');
        
        const winnerDisplay = document.getElementById('winner-display');
        if (winnerId === this.userId) {
            winnerDisplay.textContent = 'You Won!';
            winnerDisplay.style.color = '#ff0';
        } else {
            const winner = this.gameState.players[winnerId];
            winnerDisplay.textContent = `${winner.username} Won!`;
            winnerDisplay.style.color = '#f00';
        }
        
        const playerStats = players[this.userId];
        if (playerStats) {
            document.getElementById('kills-stat').textContent = playerStats.kills || 0;
            
            const playerIds = Object.keys(players);
            const sortedPlayers = playerIds.sort((a, b) => {
                if (players[a].alive && !players[b].alive) return -1;
                if (!players[a].alive && players[b].alive) return 1;
                return (players[b].kills || 0) - (players[a].kills || 0);
            });
            
            const position = sortedPlayers.indexOf(this.userId) + 1;
            document.getElementById('position-stat').textContent = `${position}${this.getOrdinalSuffix(position)}`;
        }
    }

    getOrdinalSuffix(i) {
        const j = i % 10;
        const k = i % 100;
        
        if (j === 1 && k !== 11) return 'st';
        if (j === 2 && k !== 12) return 'nd';
        if (j === 3 && k !== 13) return 'rd';
        return 'th';
    }
}

document.addEventListener('DOMContentLoaded', () => {
    window.miniRoyalGame = new MiniRoyalGame();
});

document.addEventListener('visibilitychange', () => {
    if (document.hidden && window.miniRoyalGame) {
        window.miniRoyalGame.socket.send(JSON.stringify({ type: 'pause' }));
    }
});